pyro/contrib/epidemiology/sir.py

```diff
@@ -318,3 +322,153 @@ def transition_bwd(self, params, prev, curr, t):
+        # Start with no internal infections.
+        return {"S": self.population, "I": 0}
+
+    def transition_fwd(self, params, state, t):
+        R0, X, tau, rho = params
+
+        # Sample flows between compartments.
+        S2I = pyro.sample("S2I_{}".format(t),
+                          infection_dist(individual_rate=R0 / tau,
+                                         num_susceptible=state["S"],
+                                         num_infectious=state["I"] + X,
+                                         population=self.population))
+        I2R = pyro.sample("I2R_{}".format(t),
+                          dist.Binomial(state["I"], 1 / tau))
+
+        # Update compartments with flows.
+        state["S"] = state["S"] - S2I
+        state["I"] = state["I"] + S2I - I2R
+
+        # In .transition_fwd() t will always be an integer but may lie outside
+        # of [0,self.duration) when forecasting.
+        rho_t = rho[..., t] if t < self.duration else rho[..., -1]
+        data_t = self.data[t] if t < self.duration else None
+
+        # Condition on observations.
+        pyro.sample("obs_{}".format(t),
+                    dist.ExtendedBinomial(S2I, rho_t),
+                    obs=data_t)
+
+    def transition_bwd(self, params, prev, curr, t):
+        R0, X, tau, rho = params
+
+        # Reverse the flow computation.
+        S2I = prev["S"] - curr["S"]
+        I2R = prev["I"] - curr["I"] + S2I
+
+        # Condition on flows between compartments.
+        pyro.sample("S2I_{}".format(t),
+                    infection_dist(individual_rate=R0 / tau,
+                                   num_susceptible=prev["S"],
+                                   num_infectious=prev["I"] + X,
+                                   population=self.population),
+                    obs=S2I)
+        pyro.sample("I2R_{}".format(t),
+                    dist.ExtendedBinomial(prev["I"], 1 / tau),
+                    obs=I2R)
+
+        # In .transition_bwd() t may be either an integer in [0,self.duration)
+        # or may be a nonstandard slicing object like (Ellipsis, None, None);
+        # we use the Index(-)[-] helper to support indexing with nonstandard t.
+        rho_t = Index(rho)[..., t]
+
+        # Condition on observations.
+        pyro.sample("obs_{}".format(t),
+                    dist.ExtendedBinomial(S2I, rho_t),
+                    obs=self.data[t])
+
+    def predict(self, forecast=0):
+        """
+        Augments
+        :meth:`~pyro.contrib.epidemiology.compartmental.Compartmental.predict`
+        with samples of ``first_infection``.
+
+        :param int forecast: The number of time steps to forecast forward.
+        :returns: A dictionary mapping sample site name (or compartment name)
+            to a tensor whose first dimension corresponds to sample batching.
+        :rtype: dict
+        """
+        samples = super().predict(forecast)
+        samples["first_infection"] = samples["I"].cumsum(-1).eq(0).sum(-1)
```

Comment:
can you please comment on this quantity?

Reply:
done